docs/architecture/adr-020-block-size.md

```diff
@@ -0,0 +1,67 @@
+# ADR 020: Limiting txs size inside a block
+
+## Changelog
+
+13-08-2018: Initial Draft
+15-08-2018: Second version after Dev's comments
+
+## Context
+
+We currently use MaxTxs to reap txs from the mempool when proposing a block,
+but enforce MaxBytes when unmarshalling a block, so we could easily propose a
+block thats too large to be valid.
+
+We should just remove MaxTxs all together and stick with MaxBytes, and have a
+`mempool.ReapMaxBytes`.
+
+But we can't just reap BlockSize.MaxBytes, since MaxBytes is for the entire block,
+not for the txs inside the block. There's extra amino overhead + the actual
+headers on top of the actual transactions + evidence + last commit.
+
+## Proposed solution
+
+Therefore, we should
+
+1) Get rid of MaxTxs.
+2) Rename MaxTxsBytes to MaxBytes.
+3) Add MaxNumEvidences cs parameter to limit the number of evidences per block.
+
+Proposed default values are:
+
+- MaxNumEvidences - 10
+
+When we need to ReapMaxBytes from the mempool, we calculate the upper bound as follows:
+
+```
+MaxLastCommitBytes = {number of validators currently enabled} * {MaxVoteBytes}
+MaxEvidenceBytes = {MaxNumEvidences} * {MaxEvidenceBytes}
```

Comment:
i feel like we can set an internal MaxEvidenceBytes and check if evidence is available, take up to that amount, and then subtract precisely the amount of evidence we're including rather than potentially leaving a lot of room here with every block (since most blocks will have no evidence)

Reply:
In order to do that, we would have to acquire the evidence before reaping the txs, or do 2 tx reaps. I'm not opposed to either of those possibilities.

Setting `MaxEvidenceBytes` vs `MaxNumEvidences` is independent of that however. I do think `MaxNumEvidences` is the better variable in the config.

Reply:
> rather than potentially leaving a lot of room here with every block

Yeah, did not think of that. Thank you! Let's do that.

> In order to do that, we would have to acquire the evidence before reaping the txs, or do 2 tx reaps.

we can do it in one pass
https://github.com/tendermint/tendermint/blob/29a51a87a15710c27a991241df94db22dc9f072d/consensus/state.go#L957-L958